packages/runtime/container-runtime/src/summarizerHeuristics.ts

```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
             refSequenceNumber: refSequenceNumber ?? this.lastOpSequenceNumber,
             summaryTime: Date.now(),
         };
+
+        this.numSystemOpsBefore = this.numSystemOps;
+        this.numSystemOps = 0;
+
+        this.numNonSystemOpsBefore = this.numNonSystemOps;
+        this.numNonSystemOps = 0;
     }
 
     public markLastAttemptAsSuccessful() {
         this._lastSuccessfulSummary = { ...this.lastAttempt };
     }
+
+    public adjustOnFailure(): void {
+        this.numSystemOps += this.numSystemOpsBefore;
+        this.numNonSystemOps += this.numNonSystemOpsBefore;
+    }
 }
 
 /**
  * This class contains the heuristics for when to summarize.
  */
 export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
-    private readonly idleTimer: Timer;
+    private readonly idleTimer: Timer | undefined;
+    private readonly runSummarize: (reason: SummarizeReason) => void;
 
     public constructor(
         private readonly heuristicData: ISummarizeHeuristicData,
         private readonly configuration: ISummaryConfiguration,
-        private readonly trySummarize: (reason: SummarizeReason) => void,
+        trySummarize: (reason: SummarizeReason) => void,
         private readonly minOpsForAttemptOnClose = 50,
+        private readonly summarizeStrategies: ISummarizeHeuristicStrategy[] = getDefaultSummarizeHeuristicStrategies(),
+        useIdleTimerStrategy: boolean = true,
     ) {
-        this.idleTimer = new Timer(
-            this.configuration.idleTime,
-            () => this.trySummarize("idle"));
+        if (useIdleTimerStrategy) {
```

Comment:
I personally prefer not to add boolean, but rather use this.configuration.idleTime == Infinity as a way to say timer should not be used. This makes it possible to encode it all in compact way in policy object, and not add more custom code paths / classes to handle such cases.